runtime/src/mm/cpp/ObjectFactory.hpp

```diff
@@ -0,0 +1,280 @@
+            owner_.AssertCorrectUnsafe();
+        }
+
+    private:
+        friend class ObjectFactoryStorage;
+
+        void AssertCorrect() const noexcept {
+            if (root_ == nullptr) {
+                RuntimeAssert(last_ == nullptr, "last_ must be null");
+            } else {
+                RuntimeAssert(last_ != nullptr, "last_ must not be null");
+                RuntimeAssert(last_->next_ == nullptr, "last_ must not have next");
+            }
+        }
+
+        ObjectFactoryStorage& owner_; // weak
+        std::unique_ptr<Node> root_;
+        Node* last_ = nullptr;
+    };
+
+    class Iterator {
+    public:
+        Node& operator*() noexcept { return *node_; }
+
+        Iterator& operator++() noexcept {
+            previousNode_ = node_;
+            node_ = node_->next_.get();
+            return *this;
+        }
+
+        bool operator==(const Iterator& rhs) const noexcept { return node_ == rhs.node_; }
+
+        bool operator!=(const Iterator& rhs) const noexcept { return node_ != rhs.node_; }
+
+    private:
+        friend class ObjectFactoryStorage;
+
+        Iterator(Node* previousNode, Node* node) noexcept : previousNode_(previousNode), node_(node) {}
+
+        Node* previousNode_; // Kept for `Iterable::EraseAndAdvance`.
+        Node* node_;
+    };
+
+    class Iterable : private MoveOnly {
+    public:
+        explicit Iterable(ObjectFactoryStorage& owner) noexcept : owner_(owner), guard_(owner_.mutex_) {}
+
+        Iterator begin() noexcept { return Iterator(nullptr, owner_.root_.get()); }
+        Iterator end() noexcept { return Iterator(owner_.last_, nullptr); }
+
+        void EraseAndAdvance(Iterator& iterator) noexcept { iterator.node_ = owner_.EraseUnsafe(iterator.previousNode_); }
+
+    private:
+        ObjectFactoryStorage& owner_; // weak
+        std::unique_lock<SimpleMutex> guard_;
+    };
+
+    // Lock `ObjectFactoryStorage` for safe iteration.
+    Iterable Iter() noexcept { return Iterable(*this); }
+
+private:
+    // Expects `mutex_` to be held by the current thread.
+    Node* EraseUnsafe(Node* previousNode) noexcept {
+        RuntimeAssert(root_ != nullptr, "Must not be empty");
+        AssertCorrectUnsafe();
+
+        if (previousNode == nullptr) {
+            // Deleting the root.
+            auto newRoot = std::move(root_->next_);
+            root_ = std::move(newRoot);
+            if (!root_) {
+                last_ = nullptr;
+            }
+            AssertCorrectUnsafe();
+            return root_.get();
+        }
+
+        auto node = std::move(previousNode->next_);
+        previousNode->next_ = std::move(node->next_);
+        if (!previousNode->next_) {
+            last_ = previousNode;
+        }
+
+        AssertCorrectUnsafe();
+        return previousNode->next_.get();
+    }
+
+    // Expects `mutex_` to be held by the current thread.
+    void AssertCorrectUnsafe() const noexcept {
```

Comment:
It makes some sense to mark it as `ALWAYS_INLINE`.
Motivation: when compiling release binaries, it is possible that LLVM optimizes this function only after (trying) inlining it. So if it makes a decision about inlining before optimizing this to no-op, it might decide to avoid inlining.

Reply:
Good point, done.